File: pa2/List.c
#include<stdio.h>
#include<stdlib.h>
#include "List.h"

// structs --------------------------------------------------------------------

// private NodeObj type
typedef struct NodeObj{
   int data;
   struct NodeObj* next;
   struct NodeObj* previous;
} NodeObj;

// private Node type
typedef NodeObj* Node;

// private ListObj type
typedef struct ListObj{
   Node front;
   Node back;
   Node cursor;
   int length;
   int index;
} ListObj;

// Constructors-Destructors ---------------------------------------------------

// newNode()
// Returns reference to new Node object. Initializes next and data fields.
// Private.
Node newNode(int data, Node previous, Node next){
   Node N = malloc(sizeof(NodeObj));
   N->data = data;
   N->next = NULL;
   N->previous = NULL;
   return(N);
}

// freeNode()
// Frees heap memory pointed to by *pN, sets *pN to NULL.
// Private.
void freeNode(Node* pN){
   if(pN != NULL && *pN != NULL){
      free(*pN);
      *pN = NULL;
   }
}

// newList()
// Returns reference to new empty List object.
List newList(void){
   List L;
   L = malloc(sizeof(ListObj));
   L->front = L->back = NULL;
   L->length = 0;
   L->index = -1;
   return(L);
}

// freeList()
// Frees all heap memory associated with List *pL, and sets *pL to NULL.S
void freeList(List* pL){
   if(pL!=NULL && *pL!=NULL){
      Node tempFront = (*pL)->front;
      while(tempFront != NULL){
         Node head = tempFront;
         tempFront = tempFront->next;
         free(head);
      }
      free(*pL);
      *pL = NULL;
   }
}

// Access functions -----------------------------------------------------------

int length(List L){
   if(L == NULL){
      printf("List Error: called length() on null list\n");
      exit(1);
   }
   return L->length;
}

int index(List L){
   if(L == NULL){
      printf("List Error: called index() on null list\n");
      exit(1);
   }
   return L->index;
}

int front(List L){
   if(L == NULL){
      printf("List Error: called front() on null list\n");
      exit(1);
   }
   if(L->length == 0){
      printf("List Error: called front() on empty list\n");
      exit(1);
   }
   return L->front->data;
}

int back(List L){
   if(L == NULL){
      printf("List Error: called back() on null list\n");
      exit(1);
   }
   if(L->length == 0){
      printf("List Error: called back() on empty list\n");
      exit(1);
   }
   return L->back->data;
}

int get(List L){
   if(L == NULL){
      printf("List Error: called get() on null list\n");
      exit(1);
   }
   if (L->length < 1){
      printf("List Error: called get() on empty list");
      exit(1);
   }
   if (L->index < 0){
      printf("List Error: called get() while index is undefined");
      exit(1);
   }
   return L->cursor->data;
}

int equals(List A, List B){
   if(A == NULL || B == NULL){
      printf("List Error: called equals() on null list\n");
      exit(1);
   }
   if((A->length) != (B->length)){
      return 0;
   }
   Node tempA = A->front;
   Node tempB = B->front;
   while(tempA != NULL || tempB!=NULL){
      if(tempA->data == tempB->data){
         tempA = tempA->next;
         tempB = tempB->next;
      }else{
       return 0;
      }
   }
   return 1;
}

// Manipulation procedures ----------------------------------------------------
void clear(List L){
   if(L == NULL){
      printf("List Error: called clear() on null list");
      exit(1);
   }
   L->front = NULL;
   L->back = NULL;
   L->cursor = NULL;
   L->length = 0;
   L->index = -1;
}

void moveFront(List L){
   if(L == NULL){
      printf("List Error: called moveFront() on null list");
      exit(1);
   }
   if(L->length > 0){
      L->cursor = L->front;
      L->index = 0;
   }
}

void moveBack(List L){
   if(L == NULL){
      printf("List Error: called moveBack() on null list");
      exit(1);
   }
   if(L->length > 0){
      L->cursor = L->back;
      L->index = L->length-1;
   }
}

void movePrev(List L){
   if(L == NULL){
      printf("List Error: called movePrev() on null list");
      exit(1);
   }
   if(L->cursor != NULL && L->index != 0){
      L->cursor = L->cursor->previous;
      L->index--;
   }else if(L->cursor != NULL && L->index == 0){
      L->cursor = NULL;
      L->index = -1;
   }
}

void moveNext(List L){
   if(L == NULL){
      printf("List Error: called moveNext() on null list");
      exit(1);
   }
   if(L->cursor != NULL && L->index != L->length-1){
      L->cursor = L->cursor->next;
      L->index++;
   }else if(L->cursor != NULL && L->index == L->length-1){
      L->cursor = NULL;
      L->index = -1;
   }
}

void prepend(List L, int data){
   if(L == NULL){
      printf("List Error: called prepend() on null list");
      exit(1);
   }
   Node nNode = newNode(data, NULL, NULL);
   if(L->length == 0){
      L->front = nNode;
      L->back = nNode;
   }else{
      L->front->previous = nNode;
      nNode->next = L->front;
      L->front = nNode;
   }
   L->length++;
}

void append(List L, int data){
   if(L == NULL){
      printf("List Error: called append() on null list");
      exit(1);
   }
   Node nNode = newNode(data, NULL, NULL);
   if(L->length == 0){
      L->front = nNode;
      L->back = nNode;
   }else{
      L->back->next = nNode;
      nNode->previous = L->back;
      L->back = nNode;
   }
   L->length++;
}

void insertBefore(List L, int data){
        if(L == NULL){
                printf("List Error: called insertBefore() on null list");
                exit(1);
        }
        if(L->length <= 0){
                printf("List Error: called insertBefore() on empty list");
                exit(1);
        }
        if(L->index < 0){
                printf("List Error: called insertBefore() on undefined index");
                exit(1);
        }
        if(L->index > 0){
           Node nNode = newNode(data, L->cursor->previous, L->cursor);
           nNode->previous = L->cursor->previous;
           nNode->next = L->cursor;
           if(L->cursor->previous == NULL){
              L->front = nNode;
           }else
            L->front = nNode;
        L->cursor->previous->next = nNode;
        L->length++;
    }
}

void insertAfter(List L, int data){
   if(L == NULL){
      printf("List Error: called insertBefore() on null list");
          exit(1);
   }
   if(L->length <= 0){
          printf("List Error: called insertBefore() on empty list");
          exit(1);
   }
   if(L->index < 0){
          printf("List Error: called insertBefore() on undefined index");
          exit(1);
   }
          Node nNode = newNode(data, L->cursor, L->cursor->next);
          nNode->previous = L->cursor;
          nNode->next = L->cursor->next;
          if(L->cursor->next == NULL)
             L->back = nNode;
          else
             L->cursor->next->previous = nNode;
   L->cursor->next = nNode;
   L->length++;
}

void deleteFront(List L){
   if(L == NULL){
      printf("List Error: called deleteFront() on null list");
      exit(1);
   }
   if(L->length <= 0){
      printf("List Error: called deleteFront() on empty list");
      exit(1);
   }
   if(L->cursor == L->front){
      L->cursor= NULL;
      L->index = -1;
   }
   Node temp = L->front;
   L->front = L->front->next;
   L->front->previous = NULL;
   L->length--;
   freeNode(&temp);
}

void deleteBack(List L){
   if(L == NULL){
      printf("List Error: called deleteBack() on null list");
      exit(1);
   }
   if(L->length <= 0){
      printf("List Error: called deleteBack() on empty list");
      exit(1);
   }
   if(L->cursor == L->back){
      L->cursor = NULL;
      L->index = -1;
   }
   Node temp = L->back;
   L->back = L->back->previous;
   L->back->next = NULL;
   L->length--;
   freeNode(&temp);
}

void delete(List L){
   if(L == NULL){
      printf("List Error: called delete() on null list");
      exit(1);
   }
   if(L->length <= 0){
      printf("List Error: called delete() on empty list");
      exit(1);
   }
   if(L->index < 0){
      printf("List Error: called delete() on undefined index");
      exit(1);
   }
   if(L->cursor == L->front){
      deleteFront(L);
   }
   else if(L->cursor == L->back){
      deleteBack(L);
   }
   else{
      Node temp = L->cursor;
      L->cursor->previous->next = L->cursor->next;
      L->cursor->next->previous = L->cursor->previous;
      L->cursor = NULL;
      L->length--;
      L->index = -1;
      freeNode(&temp);
   }
}

// Other operations -----------------------------------------------------------

void printList(FILE* out, List L){
   Node N = NULL;
   if( L == NULL){
      printf("List Error: called printList() on null list");
      exit(1);
   }
   for(N = L->front; N != NULL; N = N->next){
      printf("%d ", N->data);
   }
}

List copyList(List L){
   List nList = newList();
   Node tempfront = L->front;
   while(tempfront != NULL){
      append(nList,tempfront->data);
      tempfront = tempfront->next;
   }
   return nList;
}

File: pa3/List.java
//-----------------------------------------------------------------------------------------
// List.java
// Erica Tom edtom 
// pa3
//-----------------------------------------------------------------------------------------

public class List{

   private class Node{
      Object data;
      Node previous;
      Node next;

      Node(Object data, Node previous, Node next){
         this.data = data;
         this.previous = previous;
         this.next = next;
      }

      // returns the string form of data
      public String toString() {
         return String.valueOf(data);
      }

      public boolean equals(Object x){
         boolean eq = false;
         Node that;
         if(x instanceof Node){
            that = (Node) x;
            eq = (this.data.equals(that.data));
         }
         return eq;
      }

   }

   private Node front;
   private Node back;
   private Node cursor;
   private int length; //length of list
   private int index;  //index of cursor

   // Constructor
   // Creates a new empty list
   List(){
      front = null;
      back = null;
      cursor = null;
      length = 0;
      index = -1;
   }

   // Access functions

   // Returns the number of elements in this List
   int length(){
      return length;
   }

   // If cursor is defined, returns the index of the cursor element, otherwise return -1
   int index(){
      return index;
   }

   // Returns front element. pre: length()>0
   Object front(){
      if (length == 0)
         throw new RuntimeException("Error: Empty List has no front element");
         return front.data;
   }

   // Returns back eleement. pre: length()>0
   Object back(){
      if (length == 0)
         throw new RuntimeException("Error: Empty List has no back element");
      return back.data;
   }
   
   // Returns cursor element. pre: length()>0, index()>=0
   Object get(){
      if (length <= 0)
         throw new RuntimeException("Error: Unable to call get() on empty list");
      if (index < 0)
         throw new RuntimeException("Error: Unable to call get() while index is undefined");
      return cursor.data;
   }

   // Returns true if this List and L are the same integer sequence
   // The cursor is ignored in both lists.
   boolean equals(List L){
      if(L.length != length){
         return false;
      }
      Node tempL = L.front;
      Node temp = front;
      while(tempL != null && temp != null){
         if(!tempL.equals(temp)){
            return false;
         }
         tempL = tempL.next;
         temp = temp.next;
      }
      return true;
   }

   // Manipulation procedures

   // Resets this List to its original empty state
   void clear(){
      front = null;
      back = null;
      cursor = null;
      length = 0;
      index = -1;
   }

    //If List is non-empty, places the cursor under the front element, otherwise do nothing
   void moveFront(){
      if (length > 0){
         cursor = front;
         index = 0;
      }
   }

   // If cursor is non-empty, places the cursor under the back element, otherwise do nothing
   void moveBack(){
      if (length > 0){
         cursor = back;
         index = length-1;
       }
   }

   //If cursor is defined and not at front, moves cursor one step toward
   //front of this List, if cursor is defined and at front, cursor becomes
   //undefined, if cursor is undefined does nothing.
   void movePrev(){
      if(cursor != null && index != 0){ //if cursor is defined and not at front
         cursor = cursor.previous;
         index--;
      }else if(cursor != null && index == 0){
         cursor = null;
         index = -1;
      }
   }

    void moveTo(int i){
     int counter;
     if(i >= 0 && i <= length()-1){
        cursor = front;
        index = 0;
        if(i == 0){
           return;
        }
        for(counter = 0; counter < i; counter++){
           cursor = cursor.next;
           index++;
        }
     }else{
        cursor = null;
        index = -1;
     }
   }

   // If cursor is defined and not at back, moves cursor one step toward
   // back of this List, if cursor is defined and at back, cursor becomes
   // undefined, if cursor is undefined does nothing.
   void moveNext(){
      if(cursor != null && index != length-1){ //if cursor is defined and not at back
         cursor = cursor.next;
         index++;
      }else if(cursor != null && index == length-1){ //if cursor is defined and at back
         cursor = null;
         index = -1;
      }
   }

   // Insert new element into this List. If List is non-empty,
   // insertion takes place before front element
   void prepend(Object data){
      Node newNode = new Node(data,null,null);
      if(length == 0){ //list is empty
         front = newNode;
         back = newNode;
      }else{   //list is not empty
         front.previous = newNode;
         newNode.next = front;
         front = newNode;
      }
   length++;
   }

   // Insert new element into this List. If List is non-empty,
   // insertion takes place after back element.
   void append(Object data){
      Node newNode = new Node(data,null,null);
      if (length == 0){
          front = newNode;
          back = newNode;
      }else{
         back.next = newNode;
         newNode.previous = back;
         back = newNode;
      }
      length++;
   }
   
    // Insert new element before cursor.
   // Pre: length()>0, index()>=0
   void insertBefore(Object data){
      if(length <= 0){
         throw new RuntimeException("Error: Unable to insertBefore() on empty list");
      }
      if(index < 0){
         throw new RuntimeException("Error: Unable to insertBefore() on undefined index");
      }
      Node newNode = new Node(data,null,null);
      newNode.previous = cursor.previous;
      newNode.next = cursor;
      if(cursor.previous == null){
         front = newNode;
      }else{
         cursor.previous.next = newNode;
      }
      cursor.previous = newNode;
      length++;
   }

   // Inserts new element after cursor.
   // Pre: length()>0, index()>=0
   void insertAfter(Object data){
      if(length <= 0){
         throw new RuntimeException("Error: Unable to insertAfter() on empty list");
      }
      if(index < 0){
         throw new RuntimeException("Error: Unable to insertBefore() on undefined index");
      }
      Node newNode = new Node(data,null,null);
      newNode.previous = cursor;
      newNode.next = cursor.next;
      if(cursor.next == null){
         back = newNode;
      }else{
         cursor.next.previous = newNode;
      }
      cursor.next = newNode;
      length++;
   }
   
   // Deletes the front element. Pre: length()>0
   void deleteFront(){
      if(length <= 0){
        throw new RuntimeException("Erorr: Unable to deleteFront() of empty list");
      }
      if (cursor == front){
         cursor = null;
         index = -1;
      }
      front = front.next;
      front.previous = null;
      length--;
   }

   // Deletes the back element. Pre: length()>0
   void deleteBack(){
      if(length <= 0){
         throw new RuntimeException("Erorr: Unable to deleteBack() of empty list");
      }
      if(length == 1){
        clear();
      }
      if(length > 1){
          back = back.previous;
          back.next = null;
          length--;
       }
   }

   // Deletes cursor element, making cursor undefined.
   // Pre: length()>0, index()>=0
   void delete(){
      if(length <= 0){
         throw new RuntimeException("Error: Unable to delete empty list");
      }
      if(index < 0){
         throw new RuntimeException("Error: Unable to delete undefined index");
      }
      if (cursor == front){
         deleteFront();
      }
      if (cursor == back){
         deleteBack();
      }
      if (cursor != back && cursor != front){
         cursor.previous.next = cursor.next;
         cursor.next = cursor.previous;
         cursor = null;
         length--;
         index = -1;
       }
   }

   // Other methods 

   // Overrides Object's toString method. Returns a String
   // representation of this List consisting of a space
   // separated sequence of integers, with front on left.
   public String toString(){
      String printList = new String();
      Node tempfront = front;
      while(tempfront != null){
         printList = printList + tempfront.data + " ";
         tempfront = tempfront.next;
      }
      return printList;
    }
 }
                                                                                  316,1         Bot
